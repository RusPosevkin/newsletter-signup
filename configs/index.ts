import { DefaultTheme } from "styled-components";

export const theme: DefaultTheme = {
  colors: {
    primary: "#111",
    secondary: "#0070f3",
    alert: '#f1fdf1'
  },
};

// time in milliseconds for showing alert message
export const ALERT_DELAY_TIMEOUT = 5000;

export const ALERT_ANIMATION_DELAY = 500;
