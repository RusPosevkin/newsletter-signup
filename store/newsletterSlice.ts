import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";

type NewsletterState = {
  isSubmitted: boolean;
  message: string;
};

export interface INewsletterState {
  newsletterState: NewsletterState
}

const initialState: INewsletterState = {
  newsletterState: {
    isSubmitted: false,
    message: ''
  }
};

export const newsletterSlice = createSlice({
  name: 'newsletter',
  initialState,
  reducers: {
    setNewsletterState: (state, action: PayloadAction<NewsletterState>) => {
      state.newsletterState = action.payload;
    },
  },
});

export const { setNewsletterState } = newsletterSlice.actions;
export const newsletterReducer = newsletterSlice.reducer;
