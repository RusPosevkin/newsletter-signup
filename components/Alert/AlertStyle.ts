import { ALERT_ANIMATION_DELAY } from "@/configs";
import styled, { keyframes } from "styled-components";

const fadeIn = keyframes`
  from {
    transform: scale(.25);
    opacity: 0;
  }

  to {
    transform: scale(1);
    opacity: 1;
  }
`;

const fadeOut = keyframes`
  from {
    transform: scale(1);
    opacity: 1;
  }

  to {
    transform: scale(.25);
    opacity: 0;
  }
`;

const animationTimeValue = `${ALERT_ANIMATION_DELAY / 1000}s`;

export const AlertContainer = styled.div<{ 'data-hidden'?: boolean; }>`
  position: fixed;
  top: 150px;
  align-items: center;
  justify-content: center;
  text-align: center;
  width: 100%;
  padding: 10;
  background-color: ${({ theme }) => theme.colors.alert};
  visibility: ${props => props['data-hidden'] ? 'hidden' : 'visible'};
  animation: ${props => props['data-hidden'] ? fadeOut : fadeIn} ${animationTimeValue} linear;
  transition: visibility ${animationTimeValue} linear;
`;
