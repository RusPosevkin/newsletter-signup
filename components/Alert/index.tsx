import React from 'react'
import AlertComponent from "./AlertStructure";
import { useAppSelector } from '@/hooks/store';

const Alert = () => {
  const newsletterState = useAppSelector((state) => state.newsletter.newsletterState);
  console.log('newsletterState: ', newsletterState);
  // const message = 'Replace this with content from Redux'

  const componentProps = {
    message: newsletterState.message,
    showAlert: newsletterState.isSubmitted
  }

  return (
    <AlertComponent {...componentProps} />
  );
}

export default Alert;
