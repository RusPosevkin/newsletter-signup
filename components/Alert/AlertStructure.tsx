import React, { FC, useEffect, useState } from 'react'

import { ALERT_ANIMATION_DELAY, ALERT_DELAY_TIMEOUT } from '@/configs';
import { useAppDispatch } from '@/hooks/store';
import { resetNewsletterState } from '@/store/newsletterSlice';
import { AlertContainer } from './AlertStyle';

const AlertStructure: FC<IAlert> = ({ message, showAlert }) => {
  const dispatch = useAppDispatch();
  const [isHidden, setIsHidden] = useState(true);

  useEffect(() => {
    if (showAlert) {
      setIsHidden(false);

      // scheduling hiding alert
      const timeoutId = setTimeout(() => {
        setIsHidden(true);
      }, ALERT_DELAY_TIMEOUT);

      // scheduling resetting state after finishing the alert's hiding animation
      const timeoutDispatchId = setTimeout(() => {
        dispatch(resetNewsletterState());
      }, ALERT_DELAY_TIMEOUT + ALERT_ANIMATION_DELAY);

      return () => {
        clearTimeout(timeoutId);
        clearTimeout(timeoutDispatchId);
      };
    } else {
      setIsHidden(true);
    }
  }, [showAlert])

  return (
    <AlertContainer data-hidden={isHidden}>
      <h1>{message}</h1>
    </AlertContainer >
  );
}

export default AlertStructure;
