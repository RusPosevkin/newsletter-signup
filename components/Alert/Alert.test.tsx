import { act, render, screen, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom'
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';

import { theme } from '@/configs';
import { store } from '@/store/store';
import { setNewsletterState } from '@/store/newsletterSlice';
import Alert from '.'

const withProviders = (children: React.JSX.Element) => (
  <ThemeProvider theme={theme}>
    <Provider store={store}>{children}</Provider>
  </ThemeProvider>
);

test('it should match the snapshot and render', async () => {
  const { container } = render(
    withProviders(
      <Alert />
    )
  );

  expect(container).toMatchSnapshot();
  expect(container).toBeInTheDocument();
});

test('it should show a message after updating the store', async () => {
  render(
    withProviders(
      <Alert />
    )
  );

  act(() => {
    store.dispatch(setNewsletterState({
      message: 'test message',
      isSubmitted: true
    }))
  });

  await waitFor(() => {
    expect(screen.getByText('test message')).toBeInTheDocument();
  });
});
