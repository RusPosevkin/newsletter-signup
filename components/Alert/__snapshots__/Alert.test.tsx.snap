// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should match the snapshot and render 1`] = `
<div>
  <div
    class="sc-17dcb476-0 cRvGnj"
    data-hidden="true"
  >
    <h1 />
  </div>
</div>
`;
